### DISCUSSION TITLE
Brainstorming solutions to format string issues

### DISCUSSION THREAD

**Discussion by matchai** [starship/starship#1116] [General]
Howdy folks! 👋

I've been thinking about the work we're doing around format strings (#624, #1057), and have thought up some potential issues with the format string updates that may need discussion.

Below, feel free to start a thread for potential issues that can arise with format strings, and we can begin threaded discussions for each one.

**Comment by matchai** [comment]
### Prompt order without prefixes

Since format strings no longer have prefixes, it is not easily possible for us to remove the prefix for the first module in a consistent fashion. For instance, we remove the `in ` prefix before directories when they're the first module. This was mentioned by @jimleroyer on Discord.

Because of that behavior, when the directory is preceded by the username module, we get this behavior:

<img width=200 src="https://user-images.githubusercontent.com/4658208/79596546-c1a98a80-80ae-11ea-914c-ebfdc0853c34.png"/>

This may be a case for keeping prefixes, or at least for thinking about what the behavior should be once format strings are used.

This is also the case for the first module after a newline.

**Comment by matchai** [comment]
### Simple character changes

For use cases such as the [NerdFonts preset](https://starship.rs/presets/#nerd-font-symbols), it would require users to change every format string in their prompt configuration, which may cause problems when we update Starship. If we decide to alter a format string or variable name, users' prompts may not see the update or variables in-use may break.
	
<img width=300 src="https://user-images.githubusercontent.com/4658208/79597222-e3efd800-80af-11ea-8044-a0c4fa240b97.png"/>

**Reply by heyrict** [reply]
Option 1: Instead of using `[:cloud: $version](some style)`, making format default to `[$symbol$version]($style) ` and allowing users to configure only the symbol may solve the issue. We may want a deprecation period before all the users have migrated to `format`.

**Reply by heyrict** [reply]
Option 2, which is currently I have implemented, is to make the format string as simple as possible, but yet configurable with an `$all` variable that displays all the preconfigured variables. Users only need to write e.g. `format = '[symbol $all](style)'` instead.
Config migration can be done via a migration script.

**Reply by davidkna** [reply]
For the nerdfont preset in particular supporting includes might help. An option like `extends = ['builtin.nerdfont']` or `preset = "nerdfont"`.

**Comment by andytom** [comment]
### Conditional formatting of variables

In a few modules we have multiple pieces of information that can be displayed and how this information are displayed can depend on what other information is visible. An example of this is the AWS module where there is the `profile` and `region`. It is possible to have neither, one or both of these set. At the moment the AWS module takes the following approach, if neither are set then the module is not displayed, if one is set it will display that one, e.g `profile` or `region`, if they are both set it will display the region in `()` e.g `profile (region)`. With the new format strings it would be nice to allow more full customisation by making both the `profile` and `region` separate variables, so that it is possible to change the order they are displayed or the styling of each variable and also simplify the configuration by removing the `displayed_items` configuration key and using the format string. However the only way I can think of to keep the existing formatting is by adding a third variable that returns the current combined string but this feels like the wrong approach.

**Reply by heyrict** [reply]
Here are some solutions that @jletey posted on discord. He used `python` module, which displays `symbol version (virtualenv)` as a example.

> 1. ..create two or three format strings in the configuration, one for virtualenv, one for python version (or two for python and penv).

e.g.

```toml
[aws]
format = ":cloud: $profile$region"
region_format = " ($name)"
```

> 2. ..create only one format string but have a `[]()` only display if its associated variable is present.

e.g. 

```toml
[aws]
format = ':cloud: $profile[ \($region\)](none)'
```
This config will display `:cloud: $profile` only when `$region` is empty, otherwise `:cloud: $profile ($region)`

> 3. ..detach virtualenv from the *python* module and make it its own

e.g.

```toml
[aws]
format = ":cloud: $profile"
[aws_region]
format = ' \($region\)'
```

Also note that @andymac4182 have a hotfix that implements option 2.

And here are some ideas of mine.

4. Use some wrapper (e.g. `()`) that display segments in it only when all variables in the wrapper are not None. 

e.g. 

```toml
format = ':cloud: $profile( \($region\))'
```
This config will display `:cloud: $profile` when `$region` is empty, otherwise `:cloud: $profile ($region)`

@andymac4182 also noted that we can use some template engines

5. Use template engines with syntax like `:cloud: $profile{{ if region }} ($region){{ endif region }}`.

**Reply by andytom** [reply]
Thanks for the summary @heyrict, I knew I should have checked the discord channel before I posted this but probably a good thing to get it added here for better visibility. I guess what we need to do is to try to find a nice balance between making the configuration flexible to support most use cases and simple enough that it is easy to modify, update and debug.

**Reply by heyrict** [reply]
If simplicity is a must, I personally don't recommend using option 5 when we can write that with 2 or 4, even at the cost of some special use cases (e.g. display some segments when either one variable is not none). Combining 1, 2 (or 4) and 3 may be our best choice, but we have to determine which one to use in some specific condition.

**Reply by stku1985** [reply]
Another approach is to include a crate like https://crates.io/crates/evalexpr in the variable parsing; this would enable things like `${if bool_exp then (region) else ""}`.

This should be relatively straightforward (famous last words...).

**Reply by jimleroyer** [reply]
I was the one to bring suggestions that @heyrict brought here. Sorry to be late to the party here. 

I did separate the virtualenv functionality into its own  module from the python module in my PR by the way! That cover #3 for the Python module. 

No matter what the solution is picked, I would really favor simplicity. It would be easier to first develop basic features by trying to cover 80% most commonly hit use cases and later on more exhaustive features.

**Comment by matchai** [comment]
Given many of the above issues, it may be best to keep (or reduce) the existing configuration options and provide `format` as an alternative.

`format` will have less intelligent available from the start (e.g. formatting based on what variables are set), but will allow any aspect of the prompt to be changed.

Existing configuration options will be less extensive but will continue allowing for quick minor adjustments and will continue allowing prompt order to work and so on.

I worry that we may have not given some of these things enough thought when we were pitching format strings. 😰  

What are your thoughts?

**Reply by heyrict** [reply]
Well, IMHO I don't think these are major issues that would make format strings a bad choice.
I've been checking p10k these days, and found the prefix issue is not an issue as its directory part do not have a `in` prefix for all the time. It is also something we could work around by appending `in` to root as a suffix.

For the second issue, pertaining the existing symbols won't be a bad choice, but if you think the nerdfont.preset will break, I'd recommend using a preset key to do all these changes.

IMO conditional formatting is not something we need to avoid. We are just lacking some templates that tells us "use config like this when you want to format the module like this".

**Reply by jimleroyer** [reply]
It seems like the new format mechanism will fix current configuration issues but bring issues of its own as well. Some around fonts...

What about supporting profiles in starship? I have profiles in the new Windows Terminal and I can switch between what suits me. There could be one *text only* profile in starship setup by default and one for commonly supported emojis, one for nerdfonts... Just thinking loud here. 

I am wondering if the number of profiles could get too big. On the other hand, that would allow people to build custom profiles such as for the powerline theme and share it in a common github/repo.

**Reply by matchai** [reply]
Yep, I think that would be a great next step. Having baked-in profiles or presets would help folks find a configuration that suits them.

**Comment by matchai** [comment]
### Linebreak can't be disabled without setting the `prompt_order`

Previously, users were able to make starship a one-line prompt without setting the `prompt_order`. This can lead to users wanting a one-line prompt being unable to see new modules in their prompt after updating starship.

**Reply by heyrict** [reply]
This can be solved by using some meta variables e.g. `prompt_format = "\n${all:one_line}"` or `prompt_format = "\n$all"`

**Comment by heyrict** [comment]
### Proposal for configuration changes
Drawing from @matchai 's thoughts, I think the configuration with format string should be minimal but customizable.

1. symbol and style (e.g. `symbol = ":cloud: "` or `symbol = { value = ":cloud: ", style = "green" }` in `aws`)

   Due to historical reasons symbol that will not change based on some condition should be defined as a format string in the config so that users don't need to touch `format` when changing symbol and its style, which are the only options that light users may change.

   Most modules may define a format string like `format = "via [$symbol$version]($style) "`

1. conditional rendering of a variable

   An optional variable (can be described as `= if condition { Some(_) } else { None }`) is recommended to have its own variable and be used in combination with partial rendering.
   e.g. in `python`, the format would be `Py $version( \($venv\))`.

   A conditional variable (can be described as `= if cond_a { var_a } else if cond_b { var_b } else { var_c }`) is recommended to have one variable and multiple format strings.
   e.g. in `prompt`, there are two configs `style_success` and `style_error` that defines the style of the symbol. A `$style` variable would then represent `style_success` or `style_error` depending on the exit status.

   In places that controls the variable with an extra option in the config (e.g. `show_percentage = true` in `memory`), it is recommended to use two variables (e.g. `${memory#}` and `${memory%}`) for that purpose.

1. variables and meta-variables

   Variables should be as atomic as possible. If needed, some variants may be provided for the same variable.
  e.g. in a logging module `$level => info`, `$LEVEL => INFO`, `$L => I`, etc.

   Meta-variables can be used as a group of variables in a module that contains many variables that may or may not represent similar context.
   e.g. In `git_status`, `conflicted` `untracked` etc. have similar context `$count` that users may want to configure them once. With a meta-variable, the config will look like the following

   ```toml
   format = "([$status])"
   status_format = "$symbol$count"
   ```

   In a config with many variables that users may seldom want to change them at all, but only want to change the prefix and suffix, it would be better providing an `$all` variable that prints them all.

**Comment by heyrict** [comment]
### Handling errors in variables

Format string allows a module to manipulate variables on demand. e.g.

```rust
fn get_version() -> Option<String> {}

// Previously a variable should always compute no matter whether user want to display it
let version: String = get_version()?;
module.add_segment(Segment::new("version", &version));

// In format string, a variable would be computed on demand
let segments = StringFormatter::new(config.format)
    .map(|var| {
        "version" => get_version(),
        _ => None,
    })
    .parse(None)
```

But when it comes to error handling, it will be worse in a format string

```rust
fn get_version() -> Result<String, Error> {}

// Before, we could directly returns `None` when encountering an error
let version: String = match get_version() {
    Ok(version) => version,
    Err(error) => {
        log::error!("Error when attempting to get version: {}", error);
        return None;
    }
}

// However, in a format string, there is no way to make the module fail.
// The only way is to make the *variable* empty.
let segments = StringFormatter::new(config.format)
    .map(|var| {
        "version" => match get_version() {
            Ok(version) => Some(version),
            Err(version) => {
                log::error!("Error when attempting to get version: {}", error);
                None
            },
        },
        _ => None,
    })
    .parse(None);
```

Either computing variable that could fail before calling `StringFormatter::map` or making it available to store an error in a variable (e.g. make variable mapper type of `Fn(&str) -> Option<Result<VariableValue, Error>>`) will solve this.

**Comment by heyrict** [comment]
@matchai Do you think all these issues are solved?

Here is an overview:

- **Prompt order without prefixes**: I think there are many workarounds by use an always-render module at the beginning of a line, or simply make it a no-prefix module in the config, or to add some suffix (`root in`) to a module before an always-render module. Or do you think this feature is a must for our users?
- **Simple character changes**: Previous config will remain working wherever possible. See https://github.com/starship/starship/discussions/1116#discussioncomment-6579
- **Conditional formatting of variables**: Solved in #1138 
- **Linebreak can't be disabled**: Linebreak will be preserved in #1056 
- **Handling errors in variables**: Solved in #1148

**Reply by matchai** [reply]
I suppose the solution to add a suffix to some modules would work. I think, if done correctly, most users wouldn't notice a difference.

Yes, it seems we're nearing a point where the known issues are addressed. ✅

**Comment by trajano** [comment]
One idea I'd like to propose is to add support for global styles.  I think a low hanging fruit version of this would be to allow custom variables to be declared and used throughout the config.

For example

```toml
[stylesheet]
defaultBackground = "#101010"
default = "fg:#cccccc bg:${stylesheet.defaultBackground}"


[...]
style="${stylesheet.default}"

[...]
default = "bold fg:#cccccc bg:${stylesheet.defaultBackground}"
```

There does need to have some form of `cycle detection` to prevent excessive loops.  Or we can just have a simple `interpolationLimit = 5` so it will do at most 5 interpolations.  5 should be more than enough.
